userspace/libsinsp/test/sinsp_metrics.ut.cpp

```diff
@@ -0,0 +1,382 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+Copyright (C) 2024 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#ifdef __linux__
+
+#include <gtest/gtest.h>
+#include "sinsp_with_test_input.h"
+#include <libsinsp/test/helpers/threads_helpers.h>
+
+TEST_F(sinsp_with_test_input, sinsp_libs_metrics_collector_prometheus)
+{
+	m_inspector.set_sinsp_stats_v2_enabled();
+	// Extra call to verify that we don't fail
+	m_inspector.set_sinsp_stats_v2_enabled();
+	DEFAULT_TREE
+	auto evt = generate_random_event(p2_t1_tid);
+	ASSERT_EQ(get_field_as_string(evt, "proc.nthreads"), "3");
+
+	/* Snapshot current metrics and get the updated metrics_snapshot buffer */
+	uint32_t test_metrics_flags = (METRICS_V2_KERNEL_COUNTERS | METRICS_V2_LIBBPF_STATS | METRICS_V2_RESOURCE_UTILIZATION | METRICS_V2_STATE_COUNTERS);
+	auto libs_metrics_collector = std::make_unique<libs::metrics::libs_metrics_collector>(&m_inspector, test_metrics_flags);
+	auto prometheus_metrics_converter = std::make_unique<libs::metrics::prometheus_metrics_converter>();
```

Comment:
I don't think you need these two objects heap-allocated and stored in smart pointers. A plain object is fine.
```suggestion
	libs::metrics::libs_metrics_collector libs_metrics_collector(&m_inspector, test_metrics_flags);
	libs::metrics::prometheus_metrics_converter prometheus_metrics_converter;
```

Reply:
Could you elaborate? What you are suggesting isn't working.

Reply:
How does it break? The change I'm proposing turns the (smart) pointer into an object. Have you modified the calls to the member functions accordingly? I.e. `libs_metrics_collector->` becomes `libs_metrics_collector.` and the same goes for the other object.

Reply:
```
libsinsp/test/sinsp_metrics.ut.cpp:271:51: error: ‘class std::unique_ptr<libs::metrics::libs_metrics_collector, std::default_delete<libs::metrics::libs_metrics_collector> >’ has no member named ‘get_metrics’
  271 |         metrics_snapshot = libs_metrics_collector.get_metrics();
```

Is it very important? We typically init smart pointers with `std::make_unique` throughout the project.

Reply:
Not important, in the sense that your code is correct, but important to understand. There's no point in going through a pointer, let me go to the extreme... would you write code like this?

```c++
    auto str_ptr = std::make_unique<std::string>("hi, there");
    send_message(*str_ptr);
    std::string str("how's life?");
    send_message(str);
```

I'm provoking. Please, forgive me.
The reason for your error could be (reading from the line number in the error message) that you correctly turned all arrows `->` into dots `.`, but you went to another test case where you might have forgotten to change the similarly named objects that are still unique_ptr's.

Reply:
Fair yes I in fact had redefinitions. Since you ask `"how's life?"` -- it's pretty crazy right now 🤣